### PR TITLE
* modified Client to maintain an event type to transport map

### PR DESCRIPTION
**Pull request by kkharma**
* fixed latent bug in emission of sample metrics
* major version bump due to lwes-java upgrade that happened with v5.4.0
* upgraded lwes-java dependency to 2.2.0

### CODE REVIEW

File: src/main/java/org/mondemand/Client.java [hunk 1]
```diff
@@ -230,11 +232,17 @@ public void finalize() {
     contextStats.clear();
 
     // shutdown all the transports
-    if (transports != null) {
-      for (Transport t : transports) {
-        try {
-          t.shutdown();
-        } catch(TransportException e) {}
+    Set<Transport> seenTransports = new HashSet<Transport>();
```
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: some transports may appear more than once in map, so keep track of which ones have already been shut down.

File: src/main/java/org/mondemand/Client.java [hunk 2]
```diff
@@ -454,14 +463,31 @@ public void addTransportsFromConfigFile(String configFileName)
    * Adds a new transport to this client.
    * @param transport the transport object to add
    */
-  public synchronized void addTransport(Transport transport) {
```
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: did not see a need for this to be synchronized any longer...

File: src/main/java/org/mondemand/Client.java [hunk 3]
```diff
@@ -1198,20 +1235,25 @@ private void logReal(String name, int line, int level,
    * Since we cannot assume transports are thread-safe, we make this method synchronized.
    */
   private synchronized void dispatchLogs() {
-    if (this.messages == null || this.transports == null) return;
```
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: transports can't be null, so i removed these needless checks...

File: src/main/java/org/mondemand/Client.java [hunk 4]
```diff
@@ -1232,16 +1273,24 @@ private synchronized void dispatchStatsSamples() {
     try {
       Context[] contexts = this.contexts.values().toArray(new Context[0]);
       StatsMessage[] statsMsgs = this.stats.values().toArray(new StatsMessage[0]);
+
+      // snapshot samples map for dispatch
       SamplesMessage[] samplesMsgs = this.samples.values().toArray(new SamplesMessage[0]);
+      this.samples = new ConcurrentHashMap<String, SamplesMessage>();
```
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: swap out samples map when dispatching... we may miss some samples for increments being made to the snapshot map, but that should be fine.

File: src/main/java/org/mondemand/transport/LWESTransport.java [hunk 5]
```diff
@@ -312,8 +312,7 @@ protected int updateLwesEventForSamples(Event event, SamplesMessage msg,
         }
       }
     }
-    // clear the samples in the sampleMessage
-    msg.resetSamples();
```
Comment (kkharma) [inline src/main/java/org/mondemand/transport/LWESTransport.java]: the snapshot should be GC-ed after dispatch, so there is no need to reset the data here or elsewhere.

File: src/test/java/org/mondemand/tests/ClientTest.java [hunk 6]
```diff
@@ -319,6 +319,8 @@ public void testSamplesMessages() throws Exception {
     Client client = createClientNoTransports();
     LWESTransport localLwesTransport = new LWESTransport(InetAddress.getLocalHost(), 9292, null);
     client.addTransport(localLwesTransport);
+    // add a second transport to test stats sample resets
+    client.addTransport(localLwesTransport);
```
Comment (kkharma) [inline src/test/java/org/mondemand/tests/ClientTest.java]: this was done to cause the original code to fail this unit test.

File: src/test/java/org/mondemand/tests/ClientTest.java [hunk 7]
```diff
@@ -642,7 +644,7 @@ public void testClientAddTransportFromFile() {
       // everything valid
       mondemandConfigFile = File.createTempFile("mondemand_config_", ".tmp");
       output = new FileOutputStream(mondemandConfigFile);
-      Vector<String> addresses = new Vector<String>();
```
Comment (kkharma) [inline src/test/java/org/mondemand/tests/ClientTest.java]: what year is this?

File: src/test/java/org/mondemand/tests/ClientTest.java [hunk 8]
```diff
@@ -922,7 +927,6 @@ public void testInternalProblems() throws Exception {
     contexts.set(client, null);
     client.getContextKeys();
 
-    transports.set(client, null);
```
Comment (kkharma) [inline src/test/java/org/mondemand/tests/ClientTest.java]: i removed this and below, as it is not really testing a valid scenario in Client.

File: src/main/java/org/mondemand/Client.java [hunk 9]
```diff
@@ -454,14 +469,23 @@ public void addTransportsFromConfigFile(String configFileName)
    * Adds a new transport to this client.
    * @param transport the transport object to add
    */
-  public synchronized void addTransport(Transport transport) {
-    if (this.transports == null) {
-      this.transports = new Vector<Transport>();
+  public void addTransport(Transport transport) {
```
Comment (mohammadkolahdouzan) [inline src/main/java/org/mondemand/Client.java]: is this even being called from any where?

File: src/main/java/org/mondemand/Client.java [hunk 10]
```diff
@@ -454,14 +469,23 @@ public void addTransportsFromConfigFile(String configFileName)
...
+   * Adds a new event-specific transport to this client.
+   * @param eventType the event type for this transport
+   * @param transport the transport object to add
+   */
+  public void addTransport(EventType eventType, Transport transport) {
+    if (null == transport) {
+      return;
     }
+
+    this.transports.get(eventType).add(transport);
```
Comment (mohammadkolahdouzan) [inline src/main/java/org/mondemand/Client.java]: what if "this.transports.get(eventType)" returns null?

File: src/main/java/org/mondemand/Client.java (hunk 9 above)
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: it's a public method, so anyone using Client can call it.

File: src/main/java/org/mondemand/Client.java (hunk 10 above)
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: i initialize an empty list for each event type in the constructor, so theoretically no.

File: src/main/java/org/mondemand/Client.java [hunk 11]
```diff
@@ -1270,13 +1297,17 @@ private synchronized void dispatchContextStats()
...
-      for (Transport t : transports) {
-        try {
-          t.send(programId, statsMsgs.toArray(new StatsMessage[0]), null,
-                 contexts);
-        } catch(TransportException te) {
-          errorHandler.handleError("Error calling Transport.sendStats()", te);
+      List<Transport> transportList = transports.get(EventType.STATS);
+
+      if (transportList != null) {
+        for (Transport t : transportList) {
```
Comment (mohammadkolahdouzan) [inline src/main/java/org/mondemand/Client.java]: why not:
for (Transport t: transports.get(EventType.STATS))
and get rid of the above if statement.

File: src/test/java/org/mondemand/tests/ClientTest.java (hunk 7 above)
Comment (mohammadkolahdouzan) [inline src/test/java/org/mondemand/tests/ClientTest.java]: nothing wrong with Vector my friend ;)

**Comment by mohammadkolahdouzan** [comment]
other than the comment for for/if statements, lgtm.

File: src/main/java/org/mondemand/Client.java (hunk 11 above)
Comment (kkharma) [inline src/main/java/org/mondemand/Client.java]: oops missed that one

**Comment by nimaco** [comment]
lgtm

**Comment by kkharma** [comment]
thanks for the review guys...

**Pull request merged**